File: src/alienclaw/martians/validator.py
"""Validate a parsed MartianSpec against the brain registry."""
from __future__ import annotations
from dataclasses import dataclass
from typing import Any

from .substitution import SUBSTITUTION_TOKEN_RE
from .types import MAX_MARTIAN_SLOTS, MartianSpec, TOOL_ID_TABLE


@dataclass
class MartianValidationResult:
    valid: bool
    errors: tuple[str, ...]

    @classmethod
    def ok(cls) -> "MartianValidationResult":
        return cls(valid=True, errors=())

    @classmethod
    def fail(cls, *errors: str) -> "MartianValidationResult":
        return cls(valid=False, errors=tuple(errors))


def validate_martian(
    spec: MartianSpec,
    brain_registry: Any,  # BrainRegistry
) -> MartianValidationResult:
    """Validate a MartianSpec. Returns a result with errors (never raises)."""
    errors: list[str] = []

    if not spec.slots:
        return MartianValidationResult.fail("MartianSpec must have at least one slot.")

    # Finite-integer guard mirroring validator.ts:34-39 (PKT-534 parity).
    # bool is a subclass of int in Python, so it must be excluded explicitly.
    # NaN/±Inf are floats, caught by the isinstance(int) check; Python int has no
    # non-finite values, so no math.isfinite needed (and it would OverflowError on 10**309).
    for i, s in enumerate(spec.slots):
        si = s.slot_index
        if not isinstance(si, int) or isinstance(si, bool):
            errors.append(
                f"slot_index must be a finite integer, got {si!r} at slot {i}"
            )
    if errors:
        return MartianValidationResult.fail(*errors)

    indices = [s.slot_index for s in spec.slots]
    if len(set(indices)) != len(indices):
        errors.append(f"Duplicate slot_index values: {sorted(indices)}")
    if sorted(indices) != list(range(len(indices))):
        errors.append(
            f"slot_index values must be contiguous starting at 0. Got: {sorted(indices)}"
        )
    for s in spec.slots:
        if s.slot_index >= MAX_MARTIAN_SLOTS:
            errors.append(
                f"slot_index={s.slot_index} exceeds max 1 (only 2 parameter sections available in Packet 16)."
            )

    for s in spec.slots:
        if s.tool_name not in TOOL_ID_TABLE:
            errors.append(f"Tool '{s.tool_name}' not in TOOL_ID_TABLE.")
        if brain_registry.lookup_by_name(s.tool_name) is None:
            errors.append(f"Tool '{s.tool_name}' not in brain registry.")

    for s in spec.slots:
        if s.inputs_from is None:
            continue
        for field, template in s.inputs_from.fields.items():
            for m in SUBSTITUTION_TOKEN_RE.finditer(template):
                slot_num_str = m.group(2)
                if slot_num_str is not None:
                    ref_slot = int(slot_num_str)
                    if ref_slot >= s.slot_index:
                        errors.append(
                            f"Slot {s.slot_index} field '{field}': "
                            f"forward reference to slot[{ref_slot}] (must be < {s.slot_index})."
                        )
            remaining = SUBSTITUTION_TOKEN_RE.sub("", template)
            if "${" in remaining:
                errors.append(
                    f"Slot {s.slot_index} field '{field}': malformed substitution token in {template!r}"
                )

    if errors:
        return MartianValidationResult(valid=False, errors=tuple(errors))
    return MartianValidationResult.ok()

File: src/alienclaw/martians/validator.ts
/**
 * Validate a parsed MartianSpec.
 * Mirrors Python src/alienclaw/martians/validator.py
 */
import type { MartianSpec } from './types.js';
import { TOOL_ID_TABLE } from './types.js';

export interface MartianValidationResult {
  valid:  boolean;
  errors: string[];
}

const _SUBST_PATTERN = /\$\{(slot\[(\d+)\]\.output|campaign)\.([a-zA-Z_][a-zA-Z0-9_]*)\}/g;

/**
 * Validate a MartianSpec. Returns a result with errors (never throws).
 *
 * @param spec       The parsed MartianSpec.
 * @param toolNames  Set of tool names known to the brain registry. Tool names
 *                   in the spec that are not present here yield an error.
 */
export function validateMartian(
  spec: MartianSpec,
  toolNames: Set<string>,
): MartianValidationResult {
  const errors: string[] = [];

  if (spec.slots.length === 0) {
    return { valid: false, errors: ['MartianSpec must have at least one slot.'] };
  }

  const indices = spec.slots.map(s => s.slotIndex);
  const sortedIndices = [...indices].sort((a, b) => a - b);
  if (new Set(indices).size !== indices.length) {
    errors.push(`Duplicate slot_index values: ${JSON.stringify(sortedIndices)}`);
  }
  const expected = Array.from({ length: indices.length }, (_, k) => k);
  if (JSON.stringify(sortedIndices) !== JSON.stringify(expected)) {
    errors.push(
      `slot_index values must be contiguous starting at 0. Got: ${JSON.stringify(sortedIndices)}`
    );
  }
  for (const s of spec.slots) {
    if (s.slotIndex > 1) {
      errors.push(
        `slot_index=${s.slotIndex} exceeds max 1 (only 2 parameter sections available in Packet 16).`
      );
    }
  }

  for (const s of spec.slots) {
    if (!(s.toolName in TOOL_ID_TABLE)) {
      errors.push(`Tool '${s.toolName}' not in TOOL_ID_TABLE.`);
    }
    if (!toolNames.has(s.toolName)) {
      errors.push(`Tool '${s.toolName}' not in brain registry.`);
    }
  }

  for (const s of spec.slots) {
    if (s.inputsFrom === null) continue;
    for (const [field, template] of Object.entries(s.inputsFrom.fields)) {
      // Re-instantiate the regex so global lastIndex doesn't leak across calls.
      const pat = new RegExp(_SUBST_PATTERN.source, 'g');
      let m: RegExpExecArray | null;
      while ((m = pat.exec(template)) !== null) {
        const slotNumStr = m[2];
        if (slotNumStr !== undefined) {
          const refSlot = parseInt(slotNumStr, 10);
          if (refSlot >= s.slotIndex) {
            errors.push(
              `Slot ${s.slotIndex} field '${field}': ` +
              `forward reference to slot[${refSlot}] (must be < ${s.slotIndex}).`
            );
          }
        }
      }
      const remaining = template.replace(new RegExp(_SUBST_PATTERN.source, 'g'), '');
      if (remaining.includes('${')) {
        errors.push(
          `Slot ${s.slotIndex} field '${field}': malformed substitution token in '${template}'`
        );
      }
    }
  }

  if (errors.length > 0) return { valid: false, errors };
  return { valid: true, errors: [] };
}
